app/utils/RefererUrlValidator.scala

```diff
@@ -0,0 +1,40 @@
+/*
+ * Copyright 2021 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package utils
+
+import play.api.Configuration
+import javax.inject.Inject
+
+class RefererUrlValidator @Inject()(configuration: Configuration) {
+  lazy val allowedRefererServices = configuration.get[Seq[String]]("allowList.refererServices")
+
+  def isValid(refererUrl: String): Boolean =
+    refersToServiceOnAllowList(refererUrl) && onlyContainsAllowedCharacters(refererUrl)
+
+  private val charsNotAllowedPattern = "[^0-9A-Za-z/_?=&-]+".r //negated character class of allowed characters
+  private def onlyContainsAllowedCharacters(refererUrl: String): Boolean =
+    charsNotAllowedPattern.replaceAllIn(refererUrl, "").length == refererUrl.length
```

Comment:
Actually the id `charsNotAllowedPattern` is maybe confusing because if I'm not wrong this val includes instead the allowed chars. If that is true it seems you cannot use the `findAllIn`

Reply:
Yeah this is neater. Altered :+1: